src/coreclr/src/vm/classlayoutinfo.cpp

```diff
@@ -0,0 +1,1114 @@
+        if (!pEntryLock.DeadlockAwareAcquire())
+        {
+            DefineFullyQualifiedNameForClassW()
+            COMPlusThrow(kTypeLoadException, IDS_CANNOT_MARSHAL_RECURSIVE_DEF, GetFullyQualifiedNameForClassW(pMT));
+        }
+
+        if (pClass->GetNativeLayoutInfo() == nullptr)
+        {
+            EEClassNativeLayoutInfo* pNativeLayoutInfo;
+            CollectNativeLayoutFieldMetadataThrowing(pMT, &pNativeLayoutInfo);
+#ifdef FEATURE_HFA
+            CheckForNativeHFA(pMT, pNativeLayoutInfo);
+#endif
+#ifdef UNIX_AMD64_ABI
+            SystemVAmd64CheckForPassNativeStructInRegister(pMT, pNativeLayoutInfo);
+#endif
+            ((LayoutEEClass*)pClass)->m_nativeLayoutInfo = pNativeLayoutInfo;
+        }
+    }
+}
+
+VOID EEClassNativeLayoutInfo::CollectNativeLayoutFieldMetadataThrowing(MethodTable* pMT, PTR_EEClassNativeLayoutInfo* pNativeLayoutInfoOut)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        PRECONDITION(CheckPointer(pMT));
+        PRECONDITION(pMT->HasLayout());
+    }
+    CONTRACTL_END;
+
+    Module* pModule = pMT->GetModule();
+    // Internal interface for the NStruct being loaded.
+    IMDInternalImport* pInternalImport = pModule->GetMDImport();
+
+    LPCUTF8 szName;
+    LPCUTF8 szNamespace;
+    if (FAILED(pInternalImport->GetNameOfTypeDef(pMT->GetCl(), &szName, &szNamespace)))
+    {
+        szName = szNamespace = "Invalid TypeDef record";
+    }
+
+#ifdef _DEBUG
+    if (g_pConfig->ShouldBreakOnStructMarshalSetup(szName))
+        CONSISTENCY_CHECK_MSGF(false, ("BreakOnStructMarshalSetup: '%s' ", szName));
+#endif
+
+    MethodTable* pParentMT = pMT->GetParentMethodTable();
+
+    BOOL fHasNonTrivialParent = pParentMT &&
+        !pParentMT->IsObjectClass() &&
+        !pParentMT->IsValueTypeClass();
+
+    // Set some defaults based on the parent type of this type (if one exists).
+    _ASSERTE(!(fHasNonTrivialParent && !(pParentMT->HasLayout())));
+
+    BOOL fParentHasLayout = pParentMT && pParentMT->HasLayout();
+    UINT32 cbAdjustedParentLayoutNativeSize = 0;
+    EEClassNativeLayoutInfo const* pParentLayoutInfo = NULL;
+    if (fParentHasLayout)
+    {
+        pParentLayoutInfo = pParentMT->GetNativeLayoutInfo();
+        // Treat base class as an initial member.
+        cbAdjustedParentLayoutNativeSize = pParentLayoutInfo->GetSize();
+        // If the parent was originally a zero-sized explicit type but
+        // got bumped up to a size of 1 for compatibility reasons, then
+        // we need to remove the padding, but ONLY for inheritance situations.
+        if (pParentMT->GetLayoutInfo()->IsZeroSized()) {
+            CONSISTENCY_CHECK(cbAdjustedParentLayoutNativeSize == 1);
+            cbAdjustedParentLayoutNativeSize = 0;
+        }
+    }
+
+
+    CorNativeLinkType charSet;
+
+    if (!pMT->GetCharSet(&charSet))
+    {
+        UNREACHABLE_MSG("Metadata for this type should have already been verified at type-load time");
```

Comment:
Every call of `MethodTable::GetCharSet` has this. Move the error handling inside `MethodTable::GetCharSet` and make it just return the `CorNativeLinkType` ?